rfc/rfc-89/rfc-89.md

```diff
@@ -0,0 +1,348 @@
+|----------------------------------------|--------|-------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| hoodie.bucket.index.partition.rule.type | string | regex       | Set rule parser for expressions.                                                                                                                                                                                                                                |
+| hoodie.bucket.index.partition.expressions | string | null        | Users can use this parameter to specify expression and the corresponding bucket numbers (separated by commas).Multiple rules are separated by semicolons like `hoodie.bucket.index.partition.expressions=expression1,bucket-number1;expression2,bucket-number2` |
+| hoodie.bucket.index.num.buckets        | int    | 4           | Hudi bucket number per partition.                                                                                                                                                                                                                               |
+| hoodie.bucket.index.cache.maxSize             | int    | -1          | The maximum size of the partitionToBucketNumber cache, -1 for no limit                                                                                                                                                                                          |
+
+Note: In the future, more parsers will be provided, such as range expression parsers and hybrid parsers.
+Here, the parser is designed to be extensible, allowing users to customize parsers.
+
+#### Demo
+```sql
+CREATE TABLE hudi_table(
+  id BIGINT,
+  name STRING,
+  price DOUBLE
+)
+WITH (
+'connector' = 'hudi',
+'path' = 'file:///tmp/hudi_table',
+'table.type' = 'MERGE_ON_READ',
+'index.type' = 'BUCKET',
+'hoodie.bucket.index.partition.expressions' = '\\d{4}-(06-(01|17|18)|11-(01|10|11)),256',
+'hoodie.bucket.index.num.buckets' = '10'
+);
+```
+
+For the dates of 06-01, 06-17, 06-18 in June and 01-11, 10-11, 11-11 in
+November of each year (in the format of yyyy-MM-dd), the corresponding bucket number for the partition is 256
+
+For common partitions use 10 as partition bucket number
+
+
+#### Config Management
+
+The expression config will be persisted as files named as `<instant>.hashing_config` for current table, it stores in
+`.hoodie/.simple_hashing/config` directory and contains the following information in json format
+
+```json
+{
+    "<instant1>": {
+       "expression": "expression1,bucket-number1;expression2,bucket-number2",
+       "rule": "rule-engine",
+       "default": "default-bucket-number"
+    },
+    "<instant2>": {}
+}
```

Comment:
We usually write a new version as a seperate file instead of merging all the versions in one for multi-version concurrency control.

Reply:
here we can use lock before merging all versions.

and it will be better to have a seperate command to show current Hashing_meta for audit or other purpose